### PR TITLE
Support for QuickCheck testing of layout & SMRMap/FGMap objects

### PR DESCRIPTION
**Pull request by slfritchie**
Introduce QuickCheck, via overhaul of the cmdlet API for test interface

This branch changes the cmdlet API substantially to allow "quick
and easy" access to Corfu's main Java APIs from a non-Java language
(in this case, Erlang).  The goal is to allow QuickCheck (either
the commercial version EQC or the open source version PropEr) to
test Java APIs.

Each cmdlet must have a `public String[] main2(String[] args)`.
The `args` list is the cmdlet's command line arguments, or it's an
array of strings as received from the QuickCheck interface.
The return String[] array at index 0 must be either "OK" or "ERROR".
Subsequent indexes, if they exist, contain enough information to be
parseable by QuickCheck and also by humans (if using the original
cmdlet interface).

The helper functions `cmdlet.ok()` and `cmdlet.error()` make it easy
to create well-formed return values.

The `corfu_layout` and `corfu_smrobject` cmdlets have two new
arguments/command-thingies that they support: `reboot` and `reset`

* reboot: perform a soft reboot, emulating the object's behavior if it
  were being started by the JVM for the first time.  (This is much
  quicker than actually stopping & restarting the JVM.)  Any state on
  disk is consulted in the usual manner.

* reset: all state on disk is deleted, then the `reboot` method is
  called.  This is necessary for QuickCheck testing to force the
  sequencer, log unit, and layout servers to start from a known
  state prior to a QuickCheck test doing whatever random stuff it
  wants to do.

The `reboot` and `reset` commands performed by `corfu_layout` will
affect only the LayoutServer.

The `reboot` and `reset` commands performed by `corfu_smrobject` will
affect both the log unit and sequencer servers.

The `reboot` implementation should be examined very carefully by my
fellow code reviewers: this reboot stuff is tricksy.  According to
the QuickCheck models that I've created so far, they seem to work
correctly, and they don't appear to be leaking too many resources
(e.g. file descriptors, threads, memory, mmap()'ed VM regions,
etc.) ... but I wouldn't bet the farm (yet) that I've refactored
the reboot() code 100% correctly.

# Running QuickCheck tests (with EQC or PropEr)

See the new file src/test/erlang/README.md for a tutorial.
This doc also includes an overview of how the Java<->Erlang
communication is implemented.

# Mitigation for bugs 211 and 212

My laptop can hit the conditions for bugs 211 and/or 212 fairly
frequently, when using the prop_parallel() test evaluation: usually
at least once per minute.

To allow QuickCheck testing to continue despite these bugs, I've
added a temp hack to the test evaluation that checks if one of the
limited number of 211/212 exceptions has happend.  If yes, then the
failing test is discarded, and the framework continues with the
next test case.

# Partial code cruft that's included in this PR

I was working on the prototype/first draft of the Configuration Manager 
when I took the long side-track to work on QuickCheck testability.
The incomplete prototoype is sitting in the LayoutServer object.
I'm leaving this ConfigManager code in place for now; either I or
Zeeshan will resume work on it shortly.

---
This change is [Reviewable](https://reviewable.io/reviews/corfudb/corfudb/214)

### CODE REVIEW

**Comment by no2chem** [comment]
Ok, looking through this PR, but is there any reason you renamed main to main2 in ICmdlet? I think it could be renamed "main"....

**Comment by slfritchie** [comment]
I'd thought it'd be useful to use a name separate from CmdletRouter::main's name.  Not a very strongly held belief.

**Comment by slfritchie** [comment]
@no2chem main2() has been renamed in commit db4b58a.  I also discovered an OS X-related problem with the QuickCheck glue code when re-trying the README's recipe on a different machine; fixed with commit 4e5f444.

**Comment by no2chem** [comment]
Hello, this PR actually prints a bunch of "top" statements everywhere, could you remove the System.out.println() calls and submit again?

**Comment by slfritchie** [comment]
Back from vacation today ... hrm, the "bunch" was once/second reminders to finish the chain manager work, but alright.  I've pushed a commit that removes the offending statement and a few others in cmdlet-related code.  GH isn't showing that commit as part of this PR's history, nor is it allowing me to reopen the PR.  So, I'll go make a new one.

File: src/main/java/org/corfudb/infrastructure/LogUnitServer.java [hunk 1]
```diff
@@ -94,25 +101,26 @@
...
-        if ((Boolean) opts.get("--memory")) {
-            log.warn("Log unit opened in-memory mode (Maximum size={}). " +
-                    "This should be run for testing purposes only. " +
-                    "If you exceed the maximum size of the unit, old entries will be AUTOMATICALLY trimmed. " +
-                    "The unit WILL LOSE ALL DATA if it exits.", Utils.convertToByteStringRepresentation(maxCacheSize));
-            localLog = new InMemoryLog(0, Long.MAX_VALUE);
-            reset();
-        } else {
-            localLog = new RollingLog(0, Long.MAX_VALUE, logdir, (Boolean) opts.get("--sync"));
+        if (opts.get("--quickcheck-test-mode") != null &&
```
Comment (dahliamalkhi) [inline src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: same as other files: test code should not be under src/main

File: src/main/java/org/corfudb/infrastructure/LogUnitServer.java [hunk 2]
```diff
@@ -94,25 +101,26 @@
     LoadingCache<Long, LogUnitEntry> dataCache;
     long maxCacheSize;
 
-    private final AbstractLocalLog localLog;
+    private AbstractLocalLog localLog;
+
+    public static long maxLogFileSize = Integer.MAX_VALUE;  // 2GB by default
```
Comment (dahliamalkhi) [inline src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: haven't actually seen this used in this file. is this being declared as a global variable for cmdlet use?

I guess it is (saw later). Should probably turn this into a constant, upper case name.

File: src/main/java/org/corfudb/infrastructure/LogUnitServer.java [hunk 3]
```diff
@@ -173,14 +181,47 @@ public void handleMessage(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter
 
     @Override
     public void reset() {
+        String d = serverContext.getDataStore().getLogDir();
+        localLog.close();
+        if (d != null) {
+            Path dir = FileSystems.getDefault().getPath(d);
+            String prefixes[] = new String[]{"log"};
+
+            for (String pfx : prefixes) {
```
Comment (dahliamalkhi) [inline src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: again, I prefer not to see explicit pathname manipulations like thie inside the main logic of LogUnitServer

File: src/main/java/org/corfudb/infrastructure/NettyServerRouter.java [hunk 4]
```diff
@@ -71,8 +71,12 @@ public void addServer(AbstractServer server) {
      * @param outMsg Outgoing message.
      */
     public void sendResponse(ChannelHandlerContext ctx, CorfuMsg inMsg, CorfuMsg outMsg) {
+        long badEpoch = -667788;
+
+        outMsg.setEpoch(badEpoch);
         outMsg.copyBaseFields(inMsg);
-        outMsg.setEpoch(getServerEpoch());
+        if (outMsg.getEpoch() == badEpoch) { log.warn("copyBaseFields didn't clobber epoch"); System.exit(66); }
```
Comment (dahliamalkhi) [inline src/main/java/org/corfudb/infrastructure/NettyServerRouter.java]: argh. what is this?

File: src/main/java/org/corfudb/infrastructure/Rank.java [hunk 5]
```diff
@@ -30,9 +30,16 @@
      */
     @Override
     public int compareTo(Rank other) {
+        /*
```
Comment (dahliamalkhi) [inline src/main/java/org/corfudb/infrastructure/Rank.java]: please do not put such comments inside a commit PR and assume other team members will be diligent enough to remove this!

it is actually not a hack at all, it is a good safe solution, which appears in Lamport's original Paxos paper. But that's besides the point above.

File: src/main/java/org/corfudb/infrastructure/SequencerServer.java [hunk 6]
```diff
@@ -64,44 +73,44 @@ public SequencerServer(ServerContext serverContext) {
...
-                globalIndex.set(newIndex);
-            }
             log.info("Sequencer initial token set to {}", globalIndex.get());
         } catch (Exception ex) {
             log.warn("Exception parsing initial token, default to 0.", ex);
+            ex.printStackTrace();
         }
     }
 
+    private void open_fc() {
```
Comment (dahliamalkhi) [inline src/main/java/org/corfudb/infrastructure/SequencerServer.java]: Java naming convention, please.

File: src/main/java/org/corfudb/infrastructure/SequencerServer.java [hunk 7]
```diff
@@ -64,44 +73,44 @@ public SequencerServer(ServerContext serverContext) {
...
+            fc = FileChannel.open(make_checkpoint_path(),
+                    EnumSet.of(StandardOpenOption.READ, StandardOpenOption.WRITE,
+                            StandardOpenOption.CREATE, StandardOpenOption.SPARSE));
+        } catch (IOException e) {
+            log.warn("Error opening " + make_checkpoint_path() + ": " + e);
+            fc = null;
+        }
+    }
+
+    private java.nio.file.Path make_checkpoint_path() {
```
Comment (dahliamalkhi) [inline src/main/java/org/corfudb/infrastructure/SequencerServer.java]: name..

File: src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java [hunk 8]
```diff
@@ -136,6 +137,12 @@
     @Getter
     Boolean connected_p;
 
+    private Bootstrap b;
```
Comment (dahliamalkhi) [inline src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: as far as I can tell, this.b is never used, and might confuse with 'b' passed as parameter to various methods.

File: src/main/java/org/corfudb/cmdlets/corfu_as.java [hunk 9]
```diff
@@ -34,7 +39,7 @@
                     + " --version  Show version\n";
 
     @Override
-    public void main(String[] args) {
+    public String[] main(String[] args) {
```
Comment (dahliamalkhi) [inline src/main/java/org/corfudb/cmdlets/corfu_as.java]: hmm. main should not return String[].

**Comment by dahliamalkhi** [comment]
@slfritchie late review for this closed PR, please note

**Pull request merged**